### PR TITLE
Feature/song discovery

### PR DESCRIPTION
**Pull request by P-Storm**
Auto discovery, doesn't have the configuration of pattern. Default is in the PR:
`%d - %a - %t.%x` => `Dance - Artist - Track.Extension`

Discovery steps:
1. Song tags
2. `dances.json` in directory, matching on songname. (Improvement to use md5 matching because name is not stable?)
3. File name pattern matching.

### CODE REVIEW

File: Ready4Balfolk.Domain/Models/Tracks/Track.cs [hunk 1]
```diff
@@ -1,6 +1,28 @@
...
+    public bool IsValid()
+    {
+        var validDance = !string.IsNullOrWhiteSpace(Dance);
+        var validTitle = !string.IsNullOrWhiteSpace(Title);
+        var validArtist = !string.IsNullOrWhiteSpace(Artist);
+        var validDuration = Length >= TimeSpan.Zero;
+        var fileInfo = FileInfo.Exists;
+
+        return validDance && validTitle && validArtist && validDuration && fileInfo;
+    }
```
Comment (TJvL) [inline Ready4Balfolk.Domain/Models/Tracks/Track.cs:26]: FileInfo.Exists is redundant here. The only caller of IsValid() is in TrackStore.LoadDirectoryAsync, where the file was just enumerated from disk and we already successfully read tags/duration from it. If the file disappears after loading, the FileSystemWatcher handles that via OnFileDeleted.

This check also means IsValid() now hits the filesystem, turning what should be a pure data validation into an I/O operation.

File: Ready4Balfolk.Domain/Services/Tracks/Discovery/TrackInformationDiscoveryService.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            }
+
+            if (minimalSet.All(patterns.ContainsKey))
+            {
+                return patterns;
+            }
+        }
+
+        // Last effort
+        patterns.TryAdd(PatternSegment.Dance, "Not-Found");
```
Comment (TJvL) [inline Ready4Balfolk.Domain/Services/Tracks/Discovery/TrackInformationDiscoveryService.cs]: This injects the literal string "Not-Found" as a dance name, which then passes IsValid() (since it's not empty) and shows up in the UI as if it were a real dance. A sentinel value that leaks into the data model will cause confusion.

If no dance is found, it's better to let IsValid() reject the track (by leaving Dance empty), or handle the missing-dance case explicitly in the UI.

For example if the dance name is just "Dance" it would still show up and I be in the dance tree under the virtual root of the tree (Dance).

File: Ready4Balfolk.Domain/Services/Tracks/Discovery/TrackInformationDiscoveryService.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        catch (IOException)
+        {
+            throw;
+        }
+        catch (Exception exception)
+        {
+            throw new IOException(
+                $"Unable to load track duration for '{fileInfo.Name}'.", exception);
+        }
+    }
```
Comment (TJvL) [inline Ready4Balfolk.Domain/Services/Tracks/Discovery/TrackInformationDiscoveryService.cs:60]: Now that the discovery pipeline is being reworked anyway, this is a good time to clean up the exception handling here (inherited from the old TrackDiscoveryService). Wrapping all non-IOException exceptions as IOException hides the actual error type. The caller in TrackStore 
  catches FormatException or IOException, so a FormatException from TagLib gets silently wrapped and loses its identity.

```c#
private static TimeSpan GetTrackDuration(IFileInfo fileInfo)
{
    try
    {
        using var file = TagLib.File.Create(fileInfo.FullName);
        return file.Properties.Duration;                                                                                                                                                                                                                                            
    }
    catch (Exception exception) when (exception is not IOException and not FormatException)                                                                                                                                                                                         
    {                                                        
        throw new IOException($"Unable to load track duration for '{fileInfo.Name}'.", exception);
    }
} 
```

**Review by TJvL — CHANGES_REQUESTED** [review]
Good ideas I like having broader functionality to discover tracks and enable better organization of tracks.

I am not entirely sure if the json file could be useful to be honest. It's kinda unwieldy when you have thousands of tracks.

I like tags and custom string pattern a lot though 👍 

Please remember that there are some changes in this PR that are actually suppossed to be submitted in #32 and they are flagged as a review comment there. I would not want those changes to be added through here anyway. It's about loading the tracks on the UI thread.

Besides the suggested changes in-line can we maybe have some tests for corner cases as well? I would argue they are more important than happy path tests (though they don't hurt either ofcourse.)

Some cases worth adding:                                                                                                                                                                                                                                                                  
  - TrackInformationDiscoveryService: test the full pipeline -- tags found, tags partially found (fallthrough to next stage), nothing found (exception), dance only from dances.json                                                                                                        
  - TagSegmentDiscovery: files with ID3v2 tags, Xiph tags, both, neither, custom dance tag present/absent                                                                                                                                                                                   
  - FilenameSegmentDiscovery: matching pattern, non-matching pattern, custom pattern from settings       
  - PatternMatcher: special characters in filenames, regex timeout, empty segments, patterns with no tokens                                                                                                                                                                                 
  - DanceFileDiscoveryService: cache invalidation when dances.json is modified mid-session

File: Ready4Balfolk.Domain/Services/Tracks/Discovery/TrackInformationDiscoveryService.cs (hunk 2 above)
Comment (P-Storm) [reply Ready4Balfolk.Domain/Services/Tracks/Discovery/TrackInformationDiscoveryService.cs]: For now I don't require the dance. This results in less exception throwing and if not inserted it will default to "", what is not shown in the ui

File: Ready4Balfolk.Domain/Services/Tracks/Discovery/TrackInformationDiscoveryService.cs (hunk 3 above)
Comment (P-Storm) [reply Ready4Balfolk.Domain/Services/Tracks/Discovery/TrackInformationDiscoveryService.cs:60]: Cleaned this up

File: Ready4Balfolk.Domain/Services/Tracks/Discovery/TrackInformationDiscoveryService.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        Dictionary<PatternSegment, string> patterns = [];
+        foreach (var patternSegmentDiscovery in patternSegmentDiscoveries.PatternSegmentDiscoveries)
+        {
+            var discoveredPatterns = patternSegmentDiscovery.Scan(fileInfo);
+            foreach (var (segment, value) in discoveredPatterns)
+            {
+                patterns.TryAdd(segment, value);
+            }
+
+            if (minimalSet.All(patterns.ContainsKey))
```
Comment (TJvL) [inline Ready4Balfolk.Domain/Services/Tracks/Discovery/TrackInformationDiscoveryService.cs]: **Blocker:** `LoadMinimalSet` returns as soon as `Title` and `Artist` are present, and `Dance` is not part of the minimal set. Any file with ordinary Title/Artist tags — the common case — short-circuits after the tag step, so `dances.json` and filename-pattern discovery never run. The track ends up with `Dance == ""` and is then silently dropped by `TrackStore`'s `IsValid()` filter: a `Dance - Artist - Title.mp3` file that also carries normal ID3 tags vanishes from the library.

Reproduced by execution: a tag file returning Title+Artist yields only `{Artist, Title}` segments, and the resulting `Track.IsValid()` is false.

Suggestion: include `Dance` in the minimal set, or keep consulting later steps for segments that are still missing.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Ready4Balfolk.Domain/Services/Tracks/Discovery/DanceFileDiscoveryService.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        {
+            if (danceFileService.FileInfo(directoryInfo).LastWriteTimeUtc <= cached.LastWriteTimeUtc)
+            {
+                return cached.DanceEntries;
+            }
+        }
+
+        var danceEntries = danceFileService.Matches(directoryInfo);
+        var entry = new CacheEntry(directoryInfo.FullName, danceFileService.FileInfo(directoryInfo).LastWriteTimeUtc, danceEntries);
+        _cache.Add(entry.FullName, entry);
```
Comment (TJvL) [inline Ready4Balfolk.Domain/Services/Tracks/Discovery/DanceFileDiscoveryService.cs]: **Blocker:** `_cache.Add(entry.FullName, entry)` throws `ArgumentException: An item with the same key has already been added` whenever a cached entry is stale (i.e. `dances.json` was edited so its `LastWriteTimeUtc` is newer) — reproduced by execution. This needs the indexer form: `_cache[entry.FullName] = entry`.

Separately, this plain `Dictionary` is mutated from `TrackStore`'s `Parallel.ForEachAsync` (`MaxDegreeOfParallelism = 32`), so the first scan of a directory containing several untagged files races on it — duplicate-key throws plus unsynchronized-dictionary corruption. A `ConcurrentDictionary` (or a lock) is needed here.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Ready4Balfolk.Domain/Services/Tracks/Discovery/TrackInformationDiscoveryService.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            {
+                return patterns;
+            }
+        }
+
+        throw new TrackInformationDiscoveryException($"Minimal set of information not found for song: {fileInfo.FullName}");
```
Comment (TJvL) [inline Ready4Balfolk.Domain/Services/Tracks/Discovery/TrackInformationDiscoveryService.cs]: **Blocker:** the pipeline now throws exception types that `TrackStore`'s per-file filter `catch (Exception ex) when (ex is FormatException or IOException)` does not catch: `TrackInformationDiscoveryException` (thrown for every untagged, unconventionally named file), `JsonException` (malformed `dances.json`), `ArgumentException` (the cache bug above), and TagLib's `CorruptFileException` via `TagFileFactory`.

An escaped exception faults the producer task before `channel.Writer.Complete()`, so `await foreach (channel.Reader.ReadAllAsync())` in `TrackStore.LoadDirectoryAsync` never completes — one bad file leaves the whole library load hung in the loading state. The same exceptions escaping the `async void` `OnFileCreated`/`OnFileRenamed` handlers crash the process when a non-conforming file lands in a watched folder.

Suggestion: broaden/centralize the per-file exception handling, and guarantee `Writer.Complete()` runs in a `finally`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Ready4Balfolk.Domain/Services/Tracks/Discovery/Matching/PatternMatcher.cs [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    {
+        { "%a", ("artist", PatternSegment.Artist, @".+?") },
+        { "%l", ("album", PatternSegment.Album, @".+?") },
+        { "%t", ("title", PatternSegment.Title, @".+?") },
+        { "%n", ("tracknumber", PatternSegment.TrackNumber, @"\d+") },
+        { "%y", ("year", PatternSegment.Year, @"\d{4}") },
+        { "%g", ("genre", PatternSegment.Genre, @".+?") },
+
+        { "%d", ("dance", PatternSegment.Dance, @".+?") },
+        { "%x", ("extension", null, @"\w{3}") }
```
Comment (TJvL) [inline Ready4Balfolk.Domain/Services/Tracks/Discovery/Matching/PatternMatcher.cs]: **Major:** the `%x` token regex `\w{3}` cannot match the 4-character extensions `.flac` and `.aiff`, both of which `AudioFormatInformation` explicitly supports. With the default pattern, filename discovery returns nothing for such files (and combined with the minimal-set and exception issues, they get dropped or throw). Execution-verified: `Mazurka - Artist - Title.flac`/`.aiff` fail to match while `.mp3` matches. `\w{2,4}` (or similar) fixes it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Ready4Balfolk.Domain/Services/Tracks/Discovery/DanceFileService.cs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+            return [];
+        }
+
+        using var stream = fileInfo.OpenRead();
+        // Deserialize JSON from the stream
+        var entries = JsonSerializer.Deserialize<DanceFileEntry[]>(stream);
+
+        if (entries?.Length == 0)
+        {
+            WriteEmptyTemplate(directoryInfo);
```
Comment (TJvL) [inline Ready4Balfolk.Domain/Services/Tracks/Discovery/DanceFileService.cs:46]: **Major:** `WriteEmptyTemplate` is called while the `using var stream = fileInfo.OpenRead()` on the same `dances.json` (line 26) is still open, so `fileInfo.Create()` fails with `IOException: The process cannot access the file … because it is being used by another process`. Reproduced by execution under `MockFileSystem`, which enforces sharing the same way Windows does — only non-enforcing Unix file systems mask it. Dispose the read stream before rewriting the file.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Ready4Balfolk.Domain/Services/Tracks/Discovery/DanceFileService.cs [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        var fileInfo = FileInfo(directoryInfo);
+        if (!fileInfo.Exists)
+        {
+            return [];
+        }
+
+        using var stream = fileInfo.OpenRead();
+        // Deserialize JSON from the stream
+        var entries = JsonSerializer.Deserialize<DanceFileEntry[]>(stream);
```
Comment (TJvL) [inline Ready4Balfolk.Domain/Services/Tracks/Discovery/DanceFileService.cs]: **Major:** a malformed `dances.json` (it's a user-edited file) throws an unhandled `JsonException` out of `Matches` — execution-verified — and duplicate `FileName` entries throw `ArgumentException` from the `ToDictionary` on line 36. Both propagate into the scan and trigger the hang/crash described in `TrackInformationDiscoveryService`. This file should degrade gracefully: log the problem and return an empty result.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Ready4Balfolk.Domain/Models/Settings/ApplicationSettings.cs [hunk 10]
```diff
@@ -8,13 +10,15 @@ public sealed record ApplicationSettings(
     bool AutoQueueRandomTrack,
     bool AllowDuplicateTracksInQueue,
     bool RequirePlaybackConfirmation,
+    bool AdditionalSongInformationRetrieval,
```
Comment (TJvL) [inline Ready4Balfolk.Domain/Models/Settings/ApplicationSettings.cs:14]: **Minor:** `AdditionalSongInformationRetrieval` is added to settings but never read anywhere in the solution (verified by search over the full checkout), so the toggle has no effect and discovery cannot actually be disabled.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Ready4Balfolk.UI/Program.cs [hunk 11]
```diff
@@ -132,11 +135,21 @@ private static void ConfigureServices(IServiceCollection services)
...
         services.AddSingleton<IRandomTrackService, RandomTrackService>();
         services.AddSingleton<ISynonymResolutionService, SynonymResolutionService>();
 
+        // Discovery Files
+        services.AddSingleton<ITrackDiscoveryService, TrackInformationDiscoveryService>();
+        services.AddSingleton<OrderedSegmentDiscovery>();
+        services.AddSingleton<ITagFileFactory, TagFileFactory>();
+        services.AddSingleton<IPatternSegmentDiscovery, TagSegmentDiscovery>();
+        services.AddSingleton<IPatternSegmentDiscovery, DanceFileDiscovery>();
+        services.AddSingleton<IPatternSegmentDiscovery, FilenameSegmentDiscovery>(sp => new FilenameSegmentDiscovery(sp.GetRequiredService<ISettingsStore>().Current.DiscoveryPattern));
```
Comment (TJvL) [inline Ready4Balfolk.UI/Program.cs]: **Minor:** the `FilenameSegmentDiscovery` singleton captures `ISettingsStore.Current.DiscoveryPattern` once at first resolve, so later settings changes are ignored until the app restarts. This is also inconsistent with the reactive `settingsStore.Observe()` pattern used elsewhere (e.g. `App.axaml.cs`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Ready4Balfolk.Domain/Services/Tracks/Discovery/DanceFileService.cs [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+        return entries?.ToDictionary(r => r.FileName, r => r.Dance) ?? [];
+    }
+
+    private void WriteEmptyTemplate(IDirectoryInfo directoryInfo)
+    {
+        var supportedFormats = AudioFormatInformation.SupportedFormats;
+
+        var discoveredMusic = directoryInfo.EnumerateFiles()
+            .Where(r => supportedFormats.Contains(r.Extension))
```
Comment (TJvL) [inline Ready4Balfolk.Domain/Services/Tracks/Discovery/DanceFileService.cs:76]: **Minor:** `supportedFormats.Contains(r.Extension)` uses a case-sensitive `HashSet`, so e.g. `SONG.MP3` is omitted from the generated template — `main`'s extension checks (and `TrackStore.SupportedAudioFormats`) are `OrdinalIgnoreCase`. The `fileMatches.TryGetValue(fileInfo.Name, …)` lookup in `Services/DanceFileDiscovery.cs` is case-sensitive too, which misses on case-insensitive file systems.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Ready4Balfolk.Domain/Stores/Tracks/TrackStore.cs [hunk 13]
```diff
@@ -173,6 +177,11 @@ await Parallel.ForEachAsync(audioFiles,
                                     file.FullName, file.LastWriteTimeUtc, track.Length);
                             }
 
+                            if (!track.IsValid())
```
Comment (TJvL) [inline Ready4Balfolk.Domain/Stores/Tracks/TrackStore.cs]: **Minor:** tracks failing `IsValid()` are discarded with no log entry, so files disappearing from the library (see the minimal-set comment) are completely silent and undiagnosable. At minimum, log a warning with the file path.

---
_Generated by [Claude Code](https://claude.ai/code)_
